### PR TITLE
Start migrating CI to Github Actions

### PR DESCRIPTION
**Pull request by RadicalZephyr**
Templates aren't directly a feature in Github Actions, so I'm not going to attempt to parameterize the jobs the same way right now. Also, the `on:` conditions are so that the CI will fire on this PR while I'm developing it, the commented out version below is the correct final condition.

### CODE REVIEW

**Comment by RadicalZephyr** [comment]
### First step

As @carllerche suggested, I'm breaking this project up into stages.  For now, I've recreated the tests specified in `ci/azure-test-stable.yml` in github actions. The new `test_tokio.yml` file should run effectively the same set of test cases as this section of `azure_pipelines.yml`:

```yaml
# Test top level crate
- template: ci/azure-test-stable.yml
  parameters:
    name: test_tokio
    rust: stable
    displayName: Test tokio
    cross: true
    crates:
      - tokio
      - tests-integration

# Test sub crates
- template: ci/azure-test-stable.yml
  parameters:
    name: test_linux
    displayName: Test sub crates -
    rust: stable
    crates:
      - tokio-macros
      - tokio-test
      - tokio-util
      - examples
```

### Differences

Technically, it does slightly less work than the azure pipelines file since the "tokio - cargo test --features full" tests aren't run twice each on linux by the template inclusion in that second step.  Let me know if that was desirable and not in fact redundant.

There is a substantial difference in how the work is partitioned into jobs (I think). Because Github Actions doesn't have any analogs for the template inclusion or looping functionality in Azure Pipelines, I used the build matrix to run the same set of cargo invocations on the same platforms. So instead of four runners, there are 13.

#### Test Results

Test results can be seen in my [personal fork of Tokio](/RadicalZephyr/tokio/actions?query=branch%3Agithub-actions-ci)

**Comment by taiki-e** [comment]
What do you think about moving yaml file to shell script (like https://github.com/rust-lang/rust/pull/65202)?

(Personally, I think it may reduce the burden of migration (especially about the difference between syntax and supported features). In fact, the crossbeam's tests were written in shell script (but it's not as complicated a CI setting as tokio), which made the transition easier: https://github.com/crossbeam-rs/crossbeam/pull/489/files)

**Comment by RadicalZephyr** [comment]
Most of the same features are supported, and it was pretty simple to replace the usage of loop with using the job matrix.

I think `test_cross_subcrates` is really the only job that has enough commands in the body using an external script file would make sense, I'll see what it looks like to do that.

File: .github/workflows/test_tokio.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    steps:
+      - uses: actions/checkout@master
+
+      # Run `tokio` with only `full`
+      - uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --manifest-path ${{ github.workspace }}/tokio/Cargo.toml --features full
+        env:
+          RUST_BACKTRACE: 1
```
Comment (taiki-e) [inline .github/workflows/test_tokio.yml:34]: nit: probably we can move this into [the top-level env field](https://github.com/tokio-rs/tokio/pull/2531/files#diff-d3c6bb574b84b525c94de18f2617a66dR11-R13)

File: .github/workflows/test_tokio.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          CI: 'True'
```
Comment (taiki-e) [inline .github/workflows/test_tokio.yml:35]: nit: GitHub Actions always set this env, so probably this is unnecessary https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables

File: .github/workflows/test_tokio.yml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      # Run `tokio` with "unstable" cfg flag
+      - uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --manifest-path ${{ github.workspace }}/tokio/Cargo.toml --features full
+        env:
+          RUSTFLAGS: '--cfg tokio_unstable'
+          RUST_BACKTRACE: 1
+          CI: 'True'
+        name: tokio - cargo test --features full
```
Comment (taiki-e) [inline .github/workflows/test_tokio.yml:47]:
```suggestion
        name: tokio - cargo test --features full --cfg tokio_unstable
```

**Comment by taiki-e** [comment]
I merged this & I opened #2559 to track the migration of the remaining of the jobs.

**Comment by carllerche** [comment]
@RadicalZephyr thanks for getting this started and @taiki-e for following up :+1:

**Pull request merged**